File: common/src/main/java/dev/obscuria/tooltips/client/particle/TextureParticle.java
package dev.obscuria.tooltips.client.particle;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import dev.obscuria.tooltips.client.element.Transform;
import dev.obscuria.tooltips.client.renderer.ParticleData;
import dev.obscuria.tooltips.client.renderer.TooltipContext;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.resources.ResourceLocation;

public record TextureParticle(
        ResourceLocation texture,
        Transform transform
) implements TooltipParticle {

    public static final Codec<TextureParticle> CODEC;

    @Override
    public Codec<TextureParticle> codec() {
        return CODEC;
    }

    @Override
    public void render(GuiGraphics graphics, TooltipContext context, ParticleData data) {
        graphics.pose().pushPose();
        transform.apply(graphics);
        graphics.blit(texture, -8, -8, 0f, 0f, 16, 16, 16, 16);
        graphics.pose().popPose();
    }

    static {
        CODEC = RecordCodecBuilder.create(codec -> codec.group(
                ResourceLocation.CODEC.fieldOf("texture").forGetter(TextureParticle::texture),
                Transform.CODEC.optionalFieldOf("transform", Transform.DEFAULT).forGetter(TextureParticle::transform)
        ).apply(codec, TextureParticle::new));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/particle/TextureParticle.kt
package dev.obscuria.tooltips.client.particle

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.tooltips.client.renderer.TooltipContext
import dev.obscuria.tooltips.client.renderer.TooltipParticleData
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.phys.Vec3

class TextureParticle(
    val texture: ResourceLocation,
    val offset: Vec3,
    val scale: Float
) : ITooltipParticle {

    override fun codec(): Codec<TextureParticle> = CODEC

    override fun render(graphics: GuiGraphics, context: TooltipContext, data: TooltipParticleData) {
        graphics.pose().pushPose()
        graphics.pose().translate(-offset.x(), -offset.y(), -offset.z())
        graphics.pose().scale(scale, scale, scale)
        graphics.blit(texture, -8, -8, 0f, 0f, 16, 16, 16, 16)
        graphics.pose().popPose()
    }

    companion object {

        val CODEC: Codec<TextureParticle> = RecordCodecBuilder.create { codec ->
            codec.group(
                ResourceLocation.CODEC
                    .fieldOf("texture")
                    .forGetter(TextureParticle::texture),
                Vec3.CODEC
                    .optionalFieldOf("offset", Vec3.ZERO)
                    .forGetter(TextureParticle::offset),
                Codec.FLOAT
                    .optionalFieldOf("scale", 1f)
                    .forGetter(TextureParticle::scale)
            ).apply(codec, ::TextureParticle)
        }
    }
}